lib/Dialect/TTIR/Analysis/DFShardingPolicy.cpp

```diff
@@ -77,59 +78,94 @@ void DFShardingPolicy::run() {
             // currentOp output tensor shard spec, nextOp exec and nextOp output
             // tensor.
             //
-            LayoutAttr currentOpLayout = legalGrids.lookup(currentOp).front();
-            llvm::SmallVector<int64_t> currentOpShardShape =
-                currentOpLayout.getShardShape(false /*convertTileToScalar*/);
-            uint64_t l1InputUsage = currentOpLayout.getElementSizeBytes();
-            for (int64_t dim : currentOpShardShape) {
-              l1InputUsage *= dim;
-            }
-
-            LayoutAttr nextOpLayout = legalGrids.lookup(nextOp).front();
-            llvm::SmallVector<int64_t> nextOpShardShape =
-                nextOpLayout.getShardShape(false /*convertTileToScalar*/);
-            uint64_t l1OutputUsage = nextOpLayout.getElementSizeBytes();
-            for (int64_t dim : nextOpShardShape) {
-              l1OutputUsage *= dim;
-            }
+            LayoutAttr currentOpLayout = legalLayouts.lookup(currentOp).front();
+            assert(isShardedMemoryLayout(currentOpLayout.getMemLayout()));
+            llvm::ArrayRef<int64_t> currentOpOutputTensorShape =
+                mlir::cast<RankedTensorType>(currentOp->getResult(0).getType())
+                    .getShape();
+            uint64_t currentOpL1OutputUsage = deviceAttr.getLayoutSizeBytes(
+                currentOpOutputTensorShape, currentOpLayout,
+                currentOpLayout.getMemorySpace());
+
+            LayoutAttr nextOpLayout = legalLayouts.lookup(nextOp).front();
+            assert(isShardedMemoryLayout(nextOpLayout.getMemLayout()));
+            llvm::ArrayRef<int64_t> nextOpOutputTensorShape =
+                mlir::cast<RankedTensorType>(nextOp->getResult(0).getType())
+                    .getShape();
+            uint64_t nextOpL1OutputUsage = deviceAttr.getLayoutSizeBytes(
+                nextOpOutputTensorShape, nextOpLayout,
+                nextOpLayout.getMemorySpace());
 
             // Figure out this const based on exec data, but will be replaced
             // with API.
             //
             constexpr float tensorL1UsageCap = 0.8;
-            bool l1UsageValid = (l1InputUsage + l1OutputUsage) <
+            bool l1UsageValid = (currentOpL1OutputUsage + nextOpL1OutputUsage) <
                                 tensorL1UsageCap * usableL1CacheSize;
 
             if (l1UsageValid) {
-              // Add to shard chain config.
-              //
-              ShardSpec shardSpec;
-              shardSpec.op = currentOp;
-
-              // Hardcoded tensor split factor for now, until pipeline OP
-              // support is added.
+              // TODO(nobradovic)
+              // It seems that bunch of TTNN ops have constraints which prevent
+              // them from being sharded if both inputs are interleaved,
+              // so proposal for now is starting a shard chain
+              // with reshard op(at later phase only when necessary based on op
```

Comment:
@mbezuljTT We need API call for some OP which will tell us if output is sharded, if inputs need to be sharded as well(I have seen it in both matmul and binary ops as req).